### ISSUE TITLE
Add methods to access external process information (pid list, ppid, child pids, environ, cmdline, comm, exe, cwd, pid existence, signals)

### ISSUE THREAD

**Issue by ghost** [godotengine/godot-proposals#10458]
### Describe the project you are working on

Godot can already execute application processes, and read their output, etc. But I was wondering if we could potentially add a means to get foreign process information, and also send signals to foreign processes.

I have a repository below which covers such functionality in depth for a wide variety of desktop platforms:

https://github.com/time-killer-games/xproc

It currently supports Windows, macOS, Linux, Android, FreeBSD, DragonFly BSD, NetBSD, OpenBSD, Solaris, and illumos. Solaris, illumos, and (I think?) NetBSD are not yet supported in Godot, whether whole or in part. But this code covers those platforms too in case they ever get added/completed in the future, which should be convenient.

The repository also supports process execution, although Godot already has that so any code relating to process execution can be omitted from my code and what remains can be adapted to be Godot-compliant. (Things like std::string and std::vector<T> can be replaced with Godot's custom type implementation equivalents).

On Windows there are private-API's used, which were taken from Process Hacker, with a CC BY 4.0 licence which is permissive. The rest of the code I wrote from scratch mostly, while following documentation and usage examples, and it is MIT licensed as is Godot. 

The code taken from Process Hacker should keep the copyright notices in-tact, however any mention of myself (Samuel Venable) in the copyright can be removed if you like. devKathy (might) be OK with having her name removed, but I'll have to ask before I can know for sure. Although I know you guys probably don't mind giving credit. 

Additionally, the code taken from Process Hacker retrieved the private Windows API's via the developers of that software reverse-engineering Windows system libraries, I'm not sure how Godot developers feel about including reverse-engineered code in Godot. I've been told mixed stories on whether that is even legal. Although I've been lead to believe it is legal (otherwise I wouldn't have used it). Feel free to err on the side of caution. IANAL.

As for macOS, it too uses undocumented API's, although, not via reverse-engineering, but by looking at Darwin's now-archived-by-Apple source tree (as well as from the XNU kernel source tree iirc). Which means these API's could be abruptly removed in a future version of macOS without warning and could break builds and cause CI to fail out of nowhere.

Features include:

- Get the list of running Process ID's in the current user session.
- Get the Parent Process ID of a given Process ID.
- Get the list of Child Process ID's of a given Process ID.
- Get the Environment Variables of a given Process ID.
- Get the Command Line Arguments of a given Process ID.
- Get the Command Name (the Executable FileName, without the Path Component) of the given Process ID.
- Get the Executable Absolute PathName (the Executable FileName, with the Path Component) of the given Process ID.
- Get the Current Working Directory of the given Process ID.
- Search for a list of Process ID's which match the given Executable FileName, Absolute PathName, or Parent Directory.
- Search for a list of Process ID's which match the given Current Working Directory.
- Get whether a given Process ID exists in the current user session (linear search).
- Get the value of a particular Environment Variable of a given Process ID.
- Get whether a particular Environment Variable exists of a given Process ID.
- Send a Signal to a particular Process ID to force it to Terminate.
- Send a Suspend Signal to a particular Process ID.
- Send a Resume Signal to a particular Process ID.

These features in most cases are more suitable to tools than games, but you never know when there might be a game developer who actually needs this for a game in-specific. This in particular can be useful on occasion for things like headless game servers.

Let me know what you think, and I don't have enough interest to make a pull request any time soon, but if it interests you, feel free to borrow my code, adapt it, and beat me to implementing this into Godot if you want. :) I know I'd like to do it myself eventually.

Lastly, all platforms have only dependencies which are already provided by the kernel or base system. Nothing needs to be installed by the developer or end user for it to work.

Regards,
Samuel

### Describe the problem or limitation you are having in your project

Lack of means to get foreign process information.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be useful for inter-process communication, tools, headless game servers, etc.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

As mentioned already, see my repository - https://github.com/time-killer-games/xproc

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This could be added with a GDNative or GDExtension or whatever it was called, but I'd like to see it added to Godot officially.

### Is there a reason why this should be core and not an add-on in the asset library?

More people would be able to know of its existence having it officially implemented and documented, more people could benefit from it.

**Comment by AThousandShips** [comment]
You haven't explained *what* you need this for (also you put things in the wrong part so it's hard to follow what your are saying)

Please elaborate on what *problem* this solves, what use cases are you trying to solve

Proposals should be about what you need this for, I don't really see any explanation for what you or others would use this for or what project you need this for

**Comment by time-killer-games** [comment]
I honestly don't know what more there is to say that I haven't already shared. I believe I answered the questions you had before you asked. It's useful functionality, and it serves a lot of use cases, and I listed several.

If you don't like my idea you're welcome to close this, I'm not sure what more I can do.

**Comment by AThousandShips** [comment]
Then I must be missing where you wrote it, please point me to it because I re-read your proposal several times and I can't find any examples of what you would *need* this for, that's distinct from what it *can* be used for, as far as I can see you listed *features* but not what it is needed for

Please see [this page](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html) for more context

**Comment by time-killer-games** [comment]
I apologize, I'll try to explain better. Writing up a better explanation as we speak in an edit to this comment.

**Comment by time-killer-games** [comment]
I still have more I'd like to add but in regards of passing information via environment variables, while similar things can be done by passing custom command line arguments, the command line arguments can now longer be modified or updated after the child process fired. Environment variables can have their values changed at runtime in Godot, so this would allow both child and parent to read those from each other with the current (and not initial) values. One exception being with macOS however, macOS has a susctl()/KERN_PROCARGS2 limitation where it can only get the initial / startup values of the environment. But that's the only platform with that problem.

adding more context and other usages now...

**Comment by AThousandShips** [comment]
I'd suggest focusing on concrete examples of what to use it for, not what the feature itself is, i.e. "I need this to do this specific feature in a game" vs "I would like to be able to access this information"

So instead of:
* I want a way to make `AudioStreamPlayer` pick a random track

The better information is:
* I want to be able to make a system where the enemy has different sounds each time they are damaged

The important difference being the concrete case, because that gives far more useful information and opens alternatives (like my suggestions above, more able to give when concrete details were given)

**Comment by time-killer-games** [comment]
Well for reference, I wrote a FreeBSD-based OS builder script, which generates an ISO lauching games in kiosk mode, run the game on boot and shut down after process completion. A lot of people use the Raspberry Pi's to do stuff like this to create Arcade Cabients. What I would use this for in my own personal projects is a game selection screen (executable no. 1) which spawns a one of several games (each their own Godot-made exeuctable) and this would greatly simplify the process of communicating between the menu/launcher app and the games it runs modally. I'm not so sure you'll find this a good enough reason for Godot to officially support my code's features however.

**Comment by time-killer-games** [comment]
Let me know if you aren't satisfied with that explanation, i can explain at least semi-justification for the other features i listed as well if you'd like. :)

**Comment by time-killer-games** [comment]
OS.execute_with_pipe only works with stdin, stdout, and stderr. You can not use any pipes outside of that, which pipes are not exactly the same thing as what I wanted to do anyway, but being limited to those three file descriptors is still very limiting compared to my desired approach.
